enhancements/ingress/global-options-enable-hsts.md

```diff
@@ -0,0 +1,298 @@
+
+Each goal is addressed in the subsections below.
+
+### Continue to enable HSTS per-route (retain the existing feature functionality)
+The `Strict-Transport-Security` response header is currently applied per-route via an
+annotation called `haproxy.router.openshift.io/hsts_header`.  This annotation has a
+required directive `max-age`, and optional directives `includeSubDomains` and `preload`,
+with the following definitions:
+- `max-age`: delta time range in seconds during which the host is regarded as an HSTS host
+  - If set to 0, it negates the effect and the host is no longer regarded as an HSTS host
+  - `max-age` decrements over time
+- `includeSubDomains`: if present, applies to any subdomains of the host's domain name
+- `preload`: if present, tells the client to include this host in its host preload list so that
+it never needs to do an initial load to get the HSTS header (note that this is not defined in RFC 6797
+and is therefore client implementation-dependent)
+
+HSTS is currently implemented as a pattern in the HAProxy template and applied to `edge` and
+`reencrypt` routes that match the `haproxy.router.openshift.io/hsts_header` annotation:
+```gotemplate
+{{- /* hsts header in response: */}}
+ {{- $hstsOptionalTokenPattern := `(?:includeSubDomains|preload)` }}
+ {{- $hstsPattern := printf `(?:%[1]s[;])*max-age=(?:\d+|"\d+")(?:[;]%[1]s)*`  $hstsOptionalTokenPattern -}}
+...
+{{- if matchValues (print $cfg.TLSTermination) "edge" "reencrypt" }}
+    {{- with $hsts := firstMatch $hstsPattern (index $cfg.Annotations "haproxy.router.openshift.io/hsts_header") }}
+  http-response set-header Strict-Transport-Security {{$hsts}}
+    {{- end }}{{/* hsts header */}}
+  {{- end }}{{/* is "edge" or "reencrypt" */}}
+```
+#### Enable HSTS for all routes globally
+To add the header to all HTTP responses from an `edge` or `reencrypt` route, add a new router environment
+variable called `ROUTER_DEFAULT_HSTS`.  To the `IngressController` custom resource definition API defined in
+`manifests/00-custom-resource-definition.yaml`, add a new `httpHeaders` property called `hstsHeaderPolicy` to
+the `spec`.  The `hstsHeaderPolicy` should contain the text directives that configure the header, where `max-age` is
+required, and `includeSubDomains` and `preload` are optional.
+
+`ROUTER_DEFAULT_HSTS` is detected in the HAProxy configuration file processing, and can be overridden
+by any individual route's annotation.  The global `hstsHeaderPolicy` does not override an existing route
+annotation for `haproxy.router.openshift.io/hsts_header`
+
+The `hstsHeaderPolicy` variable can be configured by the administrator in the `IngressController`, as shown in this example, where
+the HSTS Policy expires after one year and includes subdomains:
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+  name: default
+  namespace: openshift-ingress-operator
+spec:
+  httpHeaders:
+    hstsHeaderPolicy: "max-age=31536000;includeSubDomains"
```

Comment:
We probably want to create a more formal API for specifying HSTS settings for an ingress controller. That way we could perform some validation on user-provided values, and ensure that all of the necessary fields are specified. The ingress-operator could then stringify the API fields into a string to be fed to the router pods via an environment variable (or maybe we could use multiple environment variables to do so). Thoughts?

Reply:
Agreed.  I was working on this today, and decided to leave this version as an alternative.